src/client/dfuse/ops/read.c

```diff
@@ -111,6 +111,333 @@ dfuse_readahead_reply(fuse_req_t req, size_t len, off_t position, struct dfuse_o
+ * be in cache.
+ *
+ * This code is entered when caching is enabled and reads are correctly size/aligned and not in the
+ * last CHUNK_SIZE of a file.  When open then the inode contains a single read_chunk_core pointer
+ * and this contains a list of read_chunk_data entries, one for each bucket.  Buckets where all
+ * slots have been requested are remove from the list and closed when the last request is completed.
+ *
+ * TODO: Currently there is no code to remove partially read buckets from the list so reading
+ * one slot every chunk would leave the entire file contents in memory until close and mean long
+ * list traversal times.
+ */
+
+#define CHUNK_SIZE (1024 * 1024)
+
+struct read_chunk_data {
+	struct dfuse_event   *ev;
+	fuse_req_t            reqs[8];
+	struct dfuse_obj_hdl *ohs[8];
+	d_list_t              list;
+	uint64_t              bucket;
+	struct dfuse_eq      *eqt;
+	int                   rc;
+	int                   entered;
+	ATOMIC int            exited;
+	bool                  exiting;
+	bool                  complete;
+};
+
+struct read_chunk_core {
+	d_list_t entries;
+};
+
+/* Global lock for all chunk read operations.  Each inode has a struct read_chunk_core * entry
+ * which is checked for NULL and set whilst holding this lock.  Each read_chunk_core then has
+ * a list of read_chunk_data and again, this lock protects all lists on all inodes.  This avoids
+ * the need for a per-inode lock which for many files would consume considerable memory but does
+ * mean there is potentially some lock contention.  The lock however is only held for list
+ * manipulation, no dfs or kernel calls are made whilst holding the lock.
+ */
+static pthread_mutex_t rc_lock = PTHREAD_MUTEX_INITIALIZER;
+
+static void
+chunk_free(struct read_chunk_data *cd)
+{
+	d_list_del(&cd->list);
+	d_slab_release(cd->eqt->de_read_slab, cd->ev);
+	D_FREE(cd);
+}
+
+/* Called when the last open file handle on a inode is closed.  This needs to free everything which
+ * is complete and for anything that isn't flag it for deletion in the callback.
+ *
+ * Returns true if the feature was used.
+ */
+bool
+read_chunk_close(struct dfuse_inode_entry *ie)
+{
+	struct read_chunk_data *cd, *cdn;
+	bool                    rcb = false;
+
+	D_MUTEX_LOCK(&rc_lock);
+	if (!ie->ie_chunk)
+		goto out;
+
+	rcb = true;
+
+	d_list_for_each_entry_safe(cd, cdn, &ie->ie_chunk->entries, list) {
+		if (cd->complete) {
+			chunk_free(cd);
+		} else {
+			cd->exiting = true;
+		}
+	}
+	D_FREE(ie->ie_chunk);
+out:
+	D_MUTEX_UNLOCK(&rc_lock);
+	return rcb;
+}
+
+static void
+chunk_cb(struct dfuse_event *ev)
+{
+	struct read_chunk_data *cd = ev->de_cd;
+	fuse_req_t              req;
+	bool                    done = false;
+
+	cd->rc = ev->de_ev.ev_error;
+
+	if (cd->rc == 0 && (ev->de_len != CHUNK_SIZE)) {
+		cd->rc = EIO;
+		DS_WARN(cd->rc, "Unexpected short read bucket %ld (%#zx) expected %i got %zi",
+			cd->bucket, cd->bucket * CHUNK_SIZE, CHUNK_SIZE, ev->de_len);
```

Comment:
`DFUSE_REPLY_ERR_RAW` be called on L254 in that case to return an error to the caller.  It needs to be done in the loop as it needs to be called for every slot where the I/O request has been entered.  For slots where the I/O request is made after the callback is complete this handled on L417 so it falls back into the standard read case there rather than returning an error.

Reply:
Make sense (it is what I was suspected).
Thanks for the clarification.